data/transactions/logic/resources.go

```diff
@@ -0,0 +1,356 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// resources contains a catalog of available resources. It's used to track the
+// apps, assets, and boxes that are available to a transaction, outside the
+// direct foreign array mechanism.
+type resources struct {
+	// These resources were created previously in the group, so they can be used
+	// by later transactions.
+	createdAsas []basics.AssetIndex
+	createdApps []basics.AppIndex
+
+	// These resources have been mentioned by some txn in the group, so they are
+	// available. But only their "main" data is available. For example, if an
+	// account is mentioned, its algo balance is available, but not necessarily
+	// its asset balance for any old ASA.
+	sharedAccounts map[basics.Address]struct{}
+	sharedAsas     map[basics.AssetIndex]struct{}
+	sharedApps     map[basics.AppIndex]struct{}
+	// We need to carefully track the "cross-product" availability, because if
+	// tx0 mentions an account A, and tx1 mentions an ASA X, that does _not_
+	// make the holding AX available
+	sharedHoldings map[ledgercore.AccountAsset]struct{}
+	sharedLocals   map[ledgercore.AccountApp]struct{}
+
+	// boxes are all of the top-level box refs from the txgroup. Most are added
+	// during NewEvalParams(). refs using 0 on an appl create are resolved and
+	// added when the appl executes. The boolean value indicates the "dirtiness"
+	// of the box - has it been modified in this txngroup? If yes, the size of
+	// the box counts against the group writeBudget. So delete is NOT a dirtying
+	// operation.
+	boxes map[boxRef]bool
+
+	// dirtyBytes maintains a running count of the number of dirty bytes in `boxes`
+	dirtyBytes uint64
+}
+
+func (r *resources) shareHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+}
+
+func (r *resources) shareAccountAndHolding(addr basics.Address, id basics.AssetIndex) {
+	r.sharedAccounts[addr] = struct{}{}
+	if id != 0 {
+		r.sharedHoldings[ledgercore.AccountAsset{Address: addr, Asset: id}] = struct{}{}
+	}
+}
+
+func (r *resources) shareLocal(addr basics.Address, id basics.AppIndex) {
+	r.sharedLocals[ledgercore.AccountApp{Address: addr, App: id}] = struct{}{}
+}
+
+// In the fill* and allows* routines, we pass the header and the fields in
+// separately, even though they are pointers into the same structure. That
+// prevents dumb attempts to use other fields from the transaction.
+
+func (r *resources) fill(tx *transactions.Transaction, ep *EvalParams) {
```

Comment:
something about how this is a very polymorphic thing to do in a language that supports polymorphic interfaces.

Reply:
Yeah, I could have made an interface like "availabilityFiller" and then had each of the transaction field types implement it. But there are two problems, one real and one maybe less so.  The real problem is that it would be a butt-ugly interface, because it would have take the widest set of arguments, even though only application calls require the ep, for example. By being explicit in a switch, each function takes what it needs.

The less real problem is that a run time conversion to an interface followed by the indirect jump is a pointless (though admittedly not very big) overhead.

Reply:
The real problem is real, the less real problem I'm not so sure is so much slower than switch plus cases.

Reply:
I wanted to benchmark, mostly out of curiosity, and now realize it would be impossible to use the interface mechanism.  This code has access to a `transactions.Transaction`, _not_ an object of a distinct type for each transaction type. The `transactions.Transaction` type _contains_ all of those other types. So, even if they each implemented some new interface, you'd need a switch on the `Type` field anyway, just to grab the right object.  It would look like this.
```
func (r *resources) fill(tx *transactions.Transaction, ep *EvalParams) {
	switch tx.Type {
	case protocol.PaymentTx:
		tx.PaymentTxnFields.fill(&tx.Header, r)
	case protocol.KeyRegistrationTx:
		tx.KeyRegistrationFields.fill(&tx.Header, r)
	case protocol.AssetConfigTx:
		// etc
	}
}
```
And, since it's doing that, there actually would be no interface usage at all, as all of the types are statically known there.  So it would just be moving the code and giving all the methods the same name. But there would never be a conversion to the interface, nor even a reason to name the "interface".

Reply:
very interesting!  thanks